### ISSUE TITLE
resolve_references() doesn't pass fromdocname parameter to the "missing-reference" event

### ISSUE THREAD

**Issue by papadokolos** [sphinx-doc/sphinx#2580]
## Background Information: (refer here if needed)

Sphinx is having troubles resolving some of the references generated from my C++ code with the help of Doxygen, breathe, and .rst files.

I found out that the reason for this is that sometimes `node["reftarget"]`'s value is missing the namespace as a prefix. e.g. node["reftarget"] == "Foo" instead of "namespace::Foo". The weird thing about it, is that it only happenes for specific parts in the document.

For example, it happenes with the `return value type` part of function's signature, but not with the same type being mentioned in the function's documentation:

> **_DocumentationPageResult.html:**_
> 
> ReturnType func(int a) `// Sphinx couldn't resolve ReturnType and it becomes a regular text`
> 
>    A member function.
> 
> **Parameters**
>     a - just a number
> 
> **Return**
>     ReturnType - some value to be returned `// Sphinx resolved a reference for ReturnType here!!`

In order to solve this problem i've written an extension which adds "namespace::" prefix the node["reftaget"] and then acts exactly as `resolve_references()` (from environment.py) does for a single iteration (since resolve_references() loops over many references).

The extension significally improves the number of resolved references (there are other issues that I haven't yet figured out), and the links were created, but they linked to a wrong file path. That leads me to the problem which im here to report on:
## The actual problem

My extension uses the function `resolve_xref()` to resolve missing references.
The function `resolve_xref()` returns a node with a link to the resolved reference, which considers the value of the "fromdocname" parameter (one of the parameters of `resolve_xref()`) - the resulted link is a relative one.

Unfortunately, my extension doesn't have "fromdocname" value to supply, because it's not included in any of the given parameters of "missing-reference" event.
The function `resolve_references()` from environment.py, which calls my extension upon a "missing-reference" event, gets "fromdocname" as a parameter, but doesn't pass it through to the event. As a result, my extension can't do it's job, the links are broken...

As a temporary solution, i've edited environemt.py as follows:

``` c++
    def resolve_references(self, doctree, fromdocname, builder):
        for node in doctree.traverse(addnodes.pending_xref):
            contnode = node[0].deepcopy()
            newnode = None

            typ = node['reftype']
            target = node['reftarget']
            refdoc = node.get('refdoc', fromdocname)
            domain = None

            try:
                ...
                # no new node found? try the missing-reference event
                if newnode is None:
                    ###################### Added this line here V (that is all)
                    node['refdoc'] = refdoc
                    ######################
                    newnode = builder.app.emit_firstresult(
                        'missing-reference', self, node, contnode)
                    # still not found? warn if node wishes to be warned about or
                    # we are in nit-picky mode
                    if newnode is None:
                        self._warn_missing_reference(refdoc, typ, target, node, domain)
            except NoUri:
                newnode = contnode
            node.replace_self(newnode or contnode)
            ...
```

Do you have a better solution in mind?

Would love to hear your thoughts,
Thank you for your time.

**Comment by jakobandersen** [comment]
Unfortunately I don't know enough about environment.py to evaluate your suggestion. However, if there is a lookup bug in the C++ domain I would very much like to fix it. The domain should keep track of namespaces in order to resolve references. Do you have an example where the problem can be reproduced? (Or is the code you mention available somewhere?)

**Comment by papadokolos** [comment]
Thank you for your quick reply!

The actual code is not available :crying_cat_face:
I'm working on an example to reproduce the problem. It will probably take me a day or two. I'll update here when it's finished.

**Comment by papadokolos** [comment]
Well, I found out that I was working with an old vesion of Sphinx (1.3b2 to be exact)...
Updating to the newest version (1.4.2) seems to fix the problem of resolving namespaces, at least in the case I've specified in the **_Background Information**_ section (the extension i've written still manages to reduce the number of warnings of unresolved references, but I didn't encounter its affect in the built `html` yet).

So, once again thank you for your willingness to assist me.
I apologize for wasting your time over such a clumsy mistake.
